stackhpc_ipa_hardware_managers/system_nic.py

```diff
@@ -0,0 +1,367 @@
+
+        :returns: HardwareSupport level for this manager.
+        """
+        # This should work for anything which supports ethtool
+        return hardware.HardwareSupport.SERVICE_PROVIDER
+
+    def get_clean_steps(self, node, ports):
+        """Get a list of clean steps with priority.
+
+        :param node: The node object as provided by Ironic.
+        :param ports: Port objects as provided by Ironic.
+        :returns: A list of cleaning steps, as a list of dicts.
+        """
+        return [{'step': 'verify_nic_firmware',
+                 'priority': 90,
+                 'interface': 'deploy',
+                 'reboot_requested': False,
+                 'abortable': True}]
+
+    def get_interface_descriptors(self):
+        """Enumerate all interfaces as NicDescriptor objects
+
+        :return: list of NicDescriptor
+        """
+        # there might be multiple identical cards, we must check them all
+        uevent_mappings = _parse_uevent(_uevent_lines())
+        result = []
+        for interface, uevent_device_info in uevent_mappings.iteritems():
+            pci_id = _get_pci_id(uevent_device_info)
+            ethtool_output = _get_ethtool_output(interface)
+            firmware = _get_ethtool_field(ethtool_output, "firmware-version")
+            result.append(NICDescriptor(pci_id=pci_id, name=interface,
+                                        firmware_version=firmware))
+        return result
+
+    def verify_nic_firmware(self, node, ports):
+        """
+        Verify all firmware versions specified by the nic_firmware property.
+
+        The nic_firmware property should take the form of a json list of
+        firmware matching criteria. The firmware matching criteria should
+        be a dictionary containing the following required fields:
+
+        - vendor_id : numeric pci vendor id
+        - device_id : numeric pci device id
+        - firmware_version : expected firmware version
+
+        the values of all fields should be strings. An example is given below::
+
+            [
+                { "vendor_id": "15B3", "device_id": "1013",
+                  "firmware_version": "12.20.1010" }
+            ]
+
+        The nic_firmware property can be manually set on a given node, as shown
+        in this example::
+
+              openstack baremetal node set $NODE_ID --extra \\
+              nic_firmware='[{"vendor_id": "15B3", "device_id": "1013", \\
+              "firmware_version": "12.20.1010"}]'
+
+        Each network card is checked against each of the items in the matching
+        criteria list. If the device and vendor ids match, the firmware version
+        is checked against the expected version specified in the
+        nic_firmware property's firmware_version field.
+
+        """
+
+        if _is_nic_verification_disabled(node):
+            LOG.warning('NIC firmware version verification has been disabled.')
+            return False
```

Comment:
Should this return True so that cleaning passes when the verification is disabled?

Reply:
I believe returning `False` will still pass:

```
Any hardware manager that returns a result from the method call is considered a success and its return value passed on to whatever dispatched the method
```

but I am returning a list of successful checks later on (so should do that here)